### PR TITLE
Followup on comments from #91612

### PR DESCRIPTION
**Pull request by idegtiarenko**
Followup on comments from #91612

### CODE REVIEW

File: server/src/test/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocatorTests.java [hunk 1]
```diff
@@ -95,93 +88,71 @@ public void testDecideShardAllocation() {
...
+            IndexMetadata.builder("light-index-1").indexWriteLoadForecast(1.0),
+            IndexMetadata.builder("light-index-2").indexWriteLoadForecast(2.0),
+            IndexMetadata.builder("light-index-3").indexWriteLoadForecast(3.0),
+            IndexMetadata.builder("zero-write-load-index").indexWriteLoadForecast(0.0),
+            IndexMetadata.builder("no-write-load-index")
+        );
 
-        var settings = Settings.EMPTY;
+        var settings = Settings.builder()
+            .put(ThrottlingAllocationDecider.CLUSTER_ROUTING_ALLOCATION_NODE_CONCURRENT_RECOVERIES_SETTING.getKey(), 10)
```
Comment (fcofdez) [inline server/src/test/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocatorTests.java:101]: Could you add a comment explaining why those settings are necessary here?

File: server/src/test/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocatorTests.java [hunk 2]
```diff
@@ -333,19 +304,23 @@ private RoutingAllocation createRoutingAllocation(ClusterState clusterState) {
...
     }
 
-    private static ClusterState stateWithStartedIndices(List<IndexMetadata.Builder> indices) {
+    private static ClusterState stateWithStartedIndices(IndexMetadata.Builder... indices) {
         var metadataBuilder = Metadata.builder();
         var routingTableBuilder = RoutingTable.builder();
         for (var index : indices) {
             var inSyncId = UUIDs.randomBase64UUID(random());
-            var build = index.putInSyncAllocationIds(0, Set.of(inSyncId)).build();
+            var build = index.settings(settings(Version.CURRENT).put("index.routing.allocation.exclude._id", "node-3"))
```
Comment (fcofdez) [inline server/src/test/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocatorTests.java:312]: is this setting used intentionally here?

File: server/src/test/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceReconcilerTests.java [hunk 3]
```diff
@@ -1024,7 +1024,6 @@ public void testDoNotRebalanceToTheNodeThatNoLongerExists() {
                     .put(SETTING_NUMBER_OF_REPLICAS, 0)
                     .put(SETTING_INDEX_VERSION_CREATED.getKey(), Version.CURRENT)
             )
-            .system(randomBoolean())
```
Comment (fcofdez) [inline server/src/test/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceReconcilerTests.java:1027]: Is this change intentional?

Comment (idegtiarenko) [reply server/src/test/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceReconcilerTests.java:1027]: Yes, this is unnecessary in that test

File: server/src/test/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocatorTests.java (hunk 2 above)
Comment (idegtiarenko) [reply server/src/test/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocatorTests.java:312]: Yes, It forces the shards to be rebalanced by a followup allocate call, I would add a comment

**Review by fcofdez — APPROVED** [review]
LGTM 👍. Thanks for adding the extra comments!

File: test/framework/src/main/java/org/elasticsearch/cluster/ESAllocationTestCase.java [hunk 4]
```diff
@@ -66,6 +71,19 @@ public Long getShardSize(ShardRouting shardRouting) {
         }
     };
 
+    public static final WriteLoadForecaster TEST_WRITE_LOAD_FORECASTER = new WriteLoadForecaster() {
```
Comment (DaveCTurner) [inline test/framework/src/main/java/org/elasticsearch/cluster/ESAllocationTestCase.java:74]: nit: this is the same as `ESAllocationTestCase#SIMULATION_WRITE_LOAD_FORECASTER` (introduced since this PR was started). I prefer the name you use here, but let's rename the existing one rather than making a new one.

File: server/src/test/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocatorTests.java [hunk 5]
```diff
@@ -95,93 +88,74 @@ public void testDecideShardAllocation() {
 
     public void testBalanceByShardLoad() {
```
Comment (DaveCTurner) [inline server/src/test/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocatorTests.java:89]: Naming nit:

```suggestion
    public void testBalanceByWriteLoad() {
```

File: server/src/test/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocatorTests.java [hunk 6]
```diff
@@ -95,93 +88,74 @@ public void testDecideShardAllocation() {
...
             for (ShardRouting shardRouting : routingNode) {
-                if (shardRouting.started() || shardRouting.initializing()) { // count load from the target node when relocating
+                if (shardRouting.started() || shardRouting.initializing()) { // count write load from the target node when relocating
                     var indexMetadata = clusterState.metadata().index(shardRouting.index());
-                    nodeIngestLoad += testWriteLoadForecaster.getForecastedWriteLoad(indexMetadata).orElse(0.0);
+                    nodeIngestLoad += TEST_WRITE_LOAD_FORECASTER.getForecastedWriteLoad(indexMetadata).orElse(0.0);
                 }
             }
-            assertThat(nodeIngestLoad, lessThanOrEqualTo(8.0 + 1.5));
+            assertThat(nodeIngestLoad, lessThanOrEqualTo(8.0));
```
Comment (DaveCTurner) [inline server/src/test/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocatorTests.java:122]: Implicitly this asserts that `heavy-index` is on a different node from `light-index-*` but could we be more explicit about that? Can we also assert that `zero-write-load-index` and `no-write-load-index` end up on the same node as `heavy-index` because of shard balance?

File: server/src/test/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocatorTests.java [hunk 7]
```diff
@@ -95,93 +88,74 @@ public void testDecideShardAllocation() {
...
             var nodeDiskUsage = 0L;
             for (ShardRouting shardRouting : routingNode) {
-                if (shardRouting.started() || shardRouting.initializing()) { // count load from the target node when relocating
+                if (shardRouting.started() || shardRouting.initializing()) { // count disk usage from the target node when relocating
                     var indexMetadata = clusterState.metadata().index(shardRouting.index());
                     nodeDiskUsage += indexMetadata.getForecastedShardSizeInBytes().orElse(0L);
                 }
             }
-            assertThat(nodeDiskUsage, lessThanOrEqualTo(ByteSizeValue.ofGb(8 + 1).getBytes()));
+            assertThat(nodeDiskUsage, lessThanOrEqualTo(ByteSizeValue.ofGb(8).getBytes()));
```
Comment (DaveCTurner) [inline server/src/test/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocatorTests.java:158]: Similarly here.

**Review by DaveCTurner — COMMENTED** [review]
I left a few small comments but overall this looks better thanks.

File: server/src/test/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocatorTests.java (hunk 5 above)
Comment (DaveCTurner) [reply server/src/test/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocatorTests.java:100]: NB ^

File: server/src/test/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocatorTests.java [hunk 8]
```diff
@@ -333,26 +338,27 @@ private RoutingAllocation createRoutingAllocation(ClusterState clusterState) {
...
+            // new ShardId(build.getIndex(), 0),
+            // // allocate indices on excluded node so that they are rebalanced according to the weights later on in test
+            // "node-3",
+            // null,
+            // true,
+            // ShardRoutingState.STARTED,
+            // AllocationId.newInitializing(inSyncId)
+            // )
+            // )
+            // );
```
Comment (DaveCTurner) [inline server/src/test/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocatorTests.java:361]: Could we remove this commented-out code?

**Review by DaveCTurner — APPROVED** [review]
LGTM except for a couple of nits
